feat(proxy): pick proxy agents by target URL's protocol to allow `http_proxy` works with https URLs and `https_proxy` works with http URLs.

fix777 opened the pull request:
## Problem Statement

`https_proxy` only works with `https` URLs, not `http` URLs.

For instance,
I have https proxy config as below, then fetch tar ball from `http://127.0.0.1:4873/jquery/-/jquery-0.0.1.tgz`.
```yaml
uplinks:
  npmjs:
    url: https://registry.npmjs.org/

packages:
  '**':
    access: $all
    publish: $authenticated
    unpublish: $authenticated
    proxy: npmjs

https_proxy: http://127.0.0.1:7890
```
- `CustomAgents` instantiated a `HttpProxyAgent` because of the `http` protocol of the `proxy` URL.
- `got` found no `HTTPS` agent, so no proxy agent used for requests to `npmjs` uplink.

## Proposed Solution

Allow `https_proxy` with `http` URLs. It's simple to setup a `http` proxy locally. So we can pick up the corresponding proxy agent by the protocol of the target request URL.

## Technical Details

Parse URL from `this.url` even if `this.proxy` exists.

```diff
class CustomAgents {
  // ...

  private getParsedUrl() {
-   return this.proxy ? new URL(this.proxy) : new URL(this.url);
+   return new URL(this.url);
  }
}
```

## Benefits

- `http_proxy` works with `https` URLs.
- `https_proxy` works with `http` URLs.

## Testing Plan

Unit tests have been added to cover `https_proxy` with `http` URLs scenarios.

juanpicado commented:
Thanks @fix777 for the patch, I will review carefully and get back to you in the next days.

fix777 commented:
@juanpicado Thanks for your time. I've read the contributors guidelines and made some changes here:

- added the required changeset summary
- changed the PR's title to match the commit guidelines

Feel free to let me know if there are any other changes needed.

juanpicado commented:
still pending on this, I haven't forgotten.

juanpicado commented on packages/proxy/src/agent.ts:
```diff
@@ -47,7 +47,7 @@ class CustomAgents {
 
   private getParsedUrl() {
-    return this.proxy ? new URL(this.proxy) : new URL(this.url);
+    return new URL(this.url);
```
I don't think this is good practice, the https_proxy variable is intended to specify a proxy server for HTTPS traffic, should not rely on the registry protocol and might damage current implementations.

fix777 replied:
Ok, get it. Can we detect protocol from `http_proxy, https_proxy` config key names? i.e. use `http_proxy` config to construct a `HttpProxyAgent` and `https_proxy` config to `HttpsProxyAgent`.

mbtools replied:
That's how it's implied in the [docs](https://verdaccio.org/docs/next/configuration/#http_proxy-and-https_proxy). Other combinations like the one you mentioned in the initial comment should lead to an error or a warning at startup of Verdaccio.

mbtools commented:
If there's no proxy, the agent is based on the protocol of the registry URL. If there's a proxy, the agent is based on the protocol of the proxy URL. If you proxy is `http`, then you need to configure is using `http_proxy`. 

Closed by #5125
